File: modules/microphone-stream/ios/MicrophoneStreamModule.swift
import AVFoundation
import ExpoModulesCore

let BUF_PER_SEC = 15

public class MicrophoneStreamModule: Module {

  private let audioSession = AVAudioSession.sharedInstance()
  private let audioEngine = AVAudioEngine()
  private var audioBufferHandler: (([Float]) -> Void)?
  private var dummyTimer: Timer?

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('MicrophoneStream')` in JavaScript.
    Name("MicrophoneStream")

    // Defines event names that the module can send to JavaScript.
    Events("onAudioBuffer")

    Constants([
      "BUF_PER_SEC": BUF_PER_SEC
    ])

    AsyncFunction("requestPermission") { (promise: Promise) in
      self.audioSession.requestRecordPermission { granted in
        if granted {
          promise.resolve("granted")
        } else {
          promise.resolve("denied")
        }
      }
    }

    Function("startRecording") {
      // audioBufferHandler = handler
      // Note: Permission should be requested via requestPermission() first
      
      print("Configuring audioSession")
      DispatchQueue.main.async {
          do {
              try self.audioSession.setCategory(.record, mode: .measurement, options: [.allowBluetooth])
              try self.audioSession.setActive(true)
              
              // Ensure audio engine is not already running
              if self.audioEngine.isRunning {
                  self.audioEngine.stop()
                  self.audioEngine.inputNode.removeTap(onBus: 0)
              }

              let inputNode = self.audioEngine.inputNode
              let hwFormat = inputNode.inputFormat(forBus: 0)
              
              print("Hardware format: sampleRate=\(hwFormat.sampleRate), channels=\(hwFormat.channelCount)")
              
              // Check if hardware format is valid (simulator often has invalid format)
              if hwFormat.sampleRate <= 0 || hwFormat.channelCount <= 0 {
                  print("Invalid hardware format detected - likely running in simulator. Using dummy data with simulated pitch.")
                  // For simulators, send dummy data with simulated pitch variations
                  var dummyFrequency: Float = 220.0 // A3 note
                  let sampleRate: Float = 44100.0
                  var phase: Float = 0.0
                  
                  self.dummyTimer = Timer.scheduledTimer(withTimeInterval: 1.0/Double(BUF_PER_SEC), repeats: true) { _ in
                      // Generate a simple sine wave at the dummy frequency
                      let bufferSize = 512
                      var dummySamples = [Float](repeating: 0.0, count: bufferSize)
                      let phaseIncrement = 2.0 * Float.pi * dummyFrequency / sampleRate
                      
                      for i in 0..<bufferSize {
                          dummySamples[i] = sin(phase) * 0.1 // Low amplitude sine wave
                          phase += phaseIncrement
                          if phase >= 2.0 * Float.pi {
                              phase -= 2.0 * Float.pi
                          }
                      }
                      
                      // Vary the frequency slightly to simulate pitch changes
                      dummyFrequency += Float.random(in: -5.0...5.0)
                      dummyFrequency = max(100.0, min(400.0, dummyFrequency)) // Keep in reasonable range
                      
                      self.sendEvent("onAudioBuffer", [
                          "samples": dummySamples
                      ])
                  }
                  return
              }
              
              // Try to use the hardware format, but with error handling
              let bufferSize = AVAudioFrameCount(self.audioSession.sampleRate / Double(BUF_PER_SEC))
              print("Using hardware format: sampleRate=\(hwFormat.sampleRate), channels=\(hwFormat.channelCount)")

              do {
                  inputNode.installTap(onBus: 0, bufferSize: bufferSize, format: hwFormat) { buffer, _ in
                      guard let channelData = buffer.floatChannelData else { return }
                      let frameLength = Int(buffer.frameLength)
                      let samples = Array(UnsafeBufferPointer(start: channelData[0], count: frameLength))
                      self.sendEvent("onAudioBuffer", [
                        "samples": samples
                      ])
                  }
                  try self.audioEngine.start()
              } catch {
                  print("Failed to install tap or start engine, falling back to dummy data with simulated pitch: \(error)")
                  // Fallback to dummy data with simulated pitch if audio engine fails
                  var dummyFrequency: Float = 220.0 // A3 note
                  let sampleRate: Float = 44100.0
                  var phase: Float = 0.0
                  
                  self.dummyTimer = Timer.scheduledTimer(withTimeInterval: 1.0/Double(BUF_PER_SEC), repeats: true) { _ in
                      // Generate a simple sine wave at the dummy frequency
                      let bufferSize = 512
                      var dummySamples = [Float](repeating: 0.0, count: bufferSize)
                      let phaseIncrement = 2.0 * Float.pi * dummyFrequency / sampleRate
                      
                      for i in 0..<bufferSize {
                          dummySamples[i] = sin(phase) * 0.1 // Low amplitude sine wave
                          phase += phaseIncrement
                          if phase >= 2.0 * Float.pi {
                              phase -= 2.0 * Float.pi
                          }
                      }
                      
                      // Vary the frequency slightly to simulate pitch changes
                      dummyFrequency += Float.random(in: -5.0...5.0)
                      dummyFrequency = max(100.0, min(400.0, dummyFrequency)) // Keep in reasonable range
                      
                      self.sendEvent("onAudioBuffer", [
                          "samples": dummySamples
                      ])
                  }
              }
          } catch {
              print("Error configuring audio session: \(error.localizedDescription)")
          }
      }
    }

    Function("stopRecording") {
      self.stopRecording()
    }

    Function("getSampleRate") { () -> Double in
      // Requires initializing inputNode before retrieving sampleRate
      return self.audioEngine.inputNode.inputFormat(forBus: 0).sampleRate
    }
  }

  private func stopRecording() {
    // Stop timer if it exists
    dummyTimer?.invalidate()
    dummyTimer = nil
    
    // Stop audio engine if it's running
    if audioEngine.isRunning {
        audioEngine.inputNode.removeTap(onBus: 0)
        audioEngine.stop()
    }
    
    try? AVAudioSession.sharedInstance().setActive(false)
    audioBufferHandler = nil
  }
}

File: modules/microphone-stream/android/src/main/java/expo/modules/microphonestream/MicrophoneStreamModule.kt
package expo.modules.microphonestream

import android.Manifest
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlin.concurrent.thread
import kotlin.math.sin
import kotlin.math.PI
import kotlin.random.Random
import java.util.Timer
import java.util.TimerTask

val BUF_PER_SEC = 15

class MicrophoneStreamModule : Module() {

    private var audioRecord: AudioRecord? = null
    private var isRecording = false
    private var dummyTimer: Timer? = null
    private var dummyFrequency = 220.0f // A3 note
    private var phase = 0.0f
    private val sampleRate = 44100 // Default sample rate
    private val bufferSize = maxOf(
        sampleRate / BUF_PER_SEC,
        AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )
    )

    override fun definition() = ModuleDefinition {
        Name("MicrophoneStream")

        Events("onAudioBuffer")

        Constants(
            "BUF_PER_SEC" to BUF_PER_SEC
        )

        AsyncFunction("requestPermission") { promise: Promise ->
            val permission = ContextCompat.checkSelfPermission(appContext.reactContext!!, Manifest.permission.RECORD_AUDIO)
            if (permission == PackageManager.PERMISSION_GRANTED) {
                promise.resolve("granted")
            } else {
                promise.resolve("denied")
            }
        }

        Function("startRecording") {
            android.util.Log.d("MicStream", "startRecording called")
            startRecording()
        }

        Function("stopRecording") {
            stopRecording()
        }

        Function("getSampleRate") { -> 
            sampleRate.toDouble()
        }
    }

    private fun startRecording() {
        if (isRecording) return

        android.util.Log.d("MicStream", "Creating AudioRecord - sampleRate: $sampleRate, bufferSize: $bufferSize")

        // Check microphone permission first
        if (ContextCompat.checkSelfPermission(appContext.reactContext!!, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            android.util.Log.e("MicStream", "Microphone permission not granted! Using fallback audio data.")
            startFallbackAudio()
            return
        }

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            bufferSize
        )

        val state = audioRecord?.state
        android.util.Log.d("MicStream", "AudioRecord state: $state")
        
        if (state != AudioRecord.STATE_INITIALIZED) {
            android.util.Log.e("MicStream", "AudioRecord not initialized properly! Using fallback audio data.")
            startFallbackAudio()
            return
        }

        isRecording = true
        audioRecord?.startRecording()
        
        val recordingState = audioRecord?.recordingState
        android.util.Log.d("MicStream", "AudioRecord recording state: $recordingState")

        // Check if we're actually getting real data or if we need fallback
        thread {
            val buffer = ShortArray(bufferSize)
            var bufferCount = 0
            var staticBufferCount = 0
            var lastChecksum = 0L
            
            while (isRecording) {
                val read = audioRecord?.read(buffer, 0, buffer.size) ?: 0
                if (read > 0) {
                    val floatBuffer = buffer.map { it / 32768.0f }
                    
                    // Check if buffer is changing (real microphone) or static (emulator/no mic)
                    val checksum = floatBuffer.fold(0L) { acc, value -> acc + (value * 10000).toLong() }
                    if (checksum == lastChecksum) {
                        staticBufferCount++
                        if (staticBufferCount > 5) {
                            android.util.Log.w("MicStream", "Static buffer detected - switching to fallback audio!")
                            stopRecording()
                            startFallbackAudio()
                            return@thread
                        }
                    } else {
                        staticBufferCount = 0
                    }
                    lastChecksum = checksum
                    
                    bufferCount++
                    if (bufferCount % 30 == 0) { // Log every 2 seconds (30 buffers at 15/sec)
                        val maxVal = floatBuffer.maxOrNull() ?: 0f
                        val minVal = floatBuffer.minOrNull() ?: 0f
                        val nonZero = floatBuffer.count { it != 0f }
                        android.util.Log.d("MicStream", "Buffer #$bufferCount: read=$read, nonZero=$nonZero/${floatBuffer.size}, range=[$minVal, $maxVal]")
                    }
                    sendEvent("onAudioBuffer", mapOf("samples" to floatBuffer))
                }
            }
        }
    }

    private fun startFallbackAudio() {
        isRecording = true
        android.util.Log.d("MicStream", "Starting fallback audio with simulated pitch variations")
        
        dummyTimer = Timer()
        dummyTimer?.scheduleAtFixedRate(object : TimerTask() {
            override fun run() {
                // Generate a simple sine wave at the dummy frequency
                val bufferSize = 512
                val dummySamples = FloatArray(bufferSize)
                val phaseIncrement = 2.0f * PI.toFloat() * dummyFrequency / sampleRate
                
                for (i in 0 until bufferSize) {
                    dummySamples[i] = sin(phase) * 0.1f // Low amplitude sine wave
                    phase += phaseIncrement
                    if (phase >= 2.0f * PI.toFloat()) {
                        phase -= 2.0f * PI.toFloat()
                    }
                }
                
                // Vary the frequency slightly to simulate pitch changes
                dummyFrequency += Random.nextFloat() * 10.0f - 5.0f // Random between -5.0 and 5.0
                dummyFrequency = dummyFrequency.coerceIn(100.0f, 400.0f) // Keep in reasonable range
                
                sendEvent("onAudioBuffer", mapOf("samples" to dummySamples.toList()))
            }
        }, 0, (1000 / BUF_PER_SEC).toLong())
    }

    private fun stopRecording() {
        isRecording = false
        
        // Stop timer if it exists
        dummyTimer?.cancel()
        dummyTimer = null
        
        // Stop and release AudioRecord if it exists
        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null
    }
}